R/vaccination.R

```diff
@@ -506,108 +506,140 @@ vaccine_schedule <- function(date, doses, n_doses = 2L) {
 ##' @title Create historical vaccine schedule
 ##'
 ##' @param data A data.frame with columns `date`, `age_band_min`,
-##'   `dose1` and `dose2`.
+##'   and numbered doses columns, e.g. if there are three doses
+##'   these should be `dose1`, `dose2` and `dose3`. Values of
+##'   `age_band_min` should be either multiples of 5 or NA - the
+##'   latter means those doses are not age-specific and will be
+##'   distributed across all ages according to priority after
+##'   all age-specific doses have already been dealt with
 ##'
-##' @param n_carehomes A vector of length 2 with the number of
-##'   carehome workers and residents.
+##' @param region Region to use to get total population numbers
+##'
+##' @param uptake A matrix of 19 rows, and number of columns equal to
+##'   number of doses. The (i,j)th entry gives the fractional uptake
+##'   of dose j for group i
 ##'
 ##' @return A [vaccine_schedule] object
 ##' @export
-vaccine_schedule_from_data <- function(data, n_carehomes) {
+vaccine_schedule_from_data <- function(data, region, uptake) {
   assert_is(data, "data.frame")
-  required <- c("age_band_min", "date", "dose1", "dose2")
+  dose_cols <- grep("dose[0-9]", names(data), value = TRUE)
+  n_doses <- length(dose_cols)
+
+  required <- c("age_band_min", "date")
   msg <- setdiff(required, names(data))
   if (length(msg) > 0) {
     stop("Required columns missing from 'data': ",
          paste(squote(msg), collapse = ", "))
   }
-  if (length(n_carehomes) != 2) {
-    stop("Expected a vector of length 2 for n_carehomes")
+
+  dose_expected <- paste0("dose", seq_len(n_doses))
+  dose_msg <- setdiff(dose_expected, names(data))
+  if (length(dose_msg) > 0) {
+    stop(sprintf("There are %s dose columns so expected dose column names: %s)",
+                 n_doses, paste(squote(dose_expected), collapse = ", ")))
   }
-  err <- is.na(data$age_band_min) | data$age_band_min %% 5 != 0
+
+  err <- data$age_band_min[!is.na(data$age_band_min)] %% 5 != 0
   if (any(err)) {
     stop("Invalid values for data$age_band_min: ",
          paste(unique(data$age_band_min[err]), collapse = ", "))
   }
+
+  age_start <- sircovid_age_bins()$start
+  if (nrow(uptake) != length(age_start) + 2) {
+    stop(sprintf("Expected uptake to have %s rows as there are %s groups",
+                 length(age_start) + 2, length(age_start) + 2))
+  }
+  if (ncol(uptake) != n_doses) {
+    stop(sprintf("Data has %s dose columns so expected uptake to have %s
+                 columns", n_doses, n_doses))
+  }
+
```

Comment:
should we also check that uptake is decreasing as dose index increases?

Reply:
good idea, have added that, plus a check that uptake is all in [0,1]